### PR TITLE
Auto Verification

### PR DESCRIPTION
**Pull request by pxwxnvermx**
[Ticket](https://dimagi.atlassian.net/browse/CCCT-290)

- Added auto-approve visits toggle to Opportunity Edit.
- Added logic to auto-approve unflagged visits and auto-reject flagged visits.
- Added logic to auto approve payments for completed and unflagged CompletedWork.

image

### CODE REVIEW

**Comment by calellowitz** [comment]
> auto-reject flagged visits

I think this is incorrect. We want to leave flagged visits as pending for the user to reject or approve.

File: commcare_connect/opportunity/visit_import.py
Comment (calellowitz) [inline commcare_connect/opportunity/visit_import.py:123]: Is the accrual logic the same here? If so you can call the function from in the task. Or just call it outside the if/else block since it happens in either case.

Relatedly, why does this need to be a separate task? The upload is already happening in a celery task.

File: commcare_connect/opportunity/views.py
Comment (calellowitz) [inline commcare_connect/opportunity/views.py:141]: I dont think you want to backfill this automatically. I think its fine to only have it affect forms going forward, but otherwise, it could be an option. I think people might accidentally do something they don't intend to if it always happens like this.

File: commcare_connect/opportunity/tasks.py
Comment (calellowitz) [inline commcare_connect/opportunity/tasks.py:284]: Even if there is no payment we want to still approve the completed work, right? In case it is necessary for other works, or just because they should show as approved, regardless of value?

On that note, does ordering matter here? So that payment units that depend on each other are approved in the right order?

File: commcare_connect/form_receiver/processor.py
Comment (calellowitz) [inline commcare_connect/form_receiver/processor.py:214]: I made this comment in the overall PR, but I think we want to auto approve only. Flags stay as pending, thats how we alert users they need to view it.

File: commcare_connect/form_receiver/processor.py
Comment (calellowitz) [inline commcare_connect/form_receiver/processor.py:217]: I would be a bit concerned about sequencing here. It may not matter but forms can be submitted many at once (since they will submit in bulk once the user gets data). We may want locking on this task, or to have it run on a schedule in bulk instead of immediately (which may make more sense anyway to do it as a batch)

**Review by calellowitz — CHANGES_REQUESTED** [review]
Left a fair number of comments, let me know if you have any questions

File: commcare_connect/opportunity/views.py
Comment (pxwxnvermx) [reply commcare_connect/opportunity/views.py]: Fixed. https://github.com/dimagi/commcare-connect/pull/294/commits/b1698ae22d0906989979e59cafd6dc51ab214630

File: commcare_connect/form_receiver/processor.py
Comment (pxwxnvermx) [reply commcare_connect/form_receiver/processor.py]: Fix. https://github.com/dimagi/commcare-connect/pull/294/commits/28a898593ab863404c5f4a79fce74f0c6d443491

File: commcare_connect/opportunity/tasks.py
Comment (pxwxnvermx) [reply commcare_connect/opportunity/tasks.py]: Fixed. https://github.com/dimagi/commcare-connect/pull/294/commits/0d34603ed9262705f8f606e05302fbc577776f21

File: commcare_connect/opportunity/visit_import.py
Comment (pxwxnvermx) [reply commcare_connect/opportunity/visit_import.py:123]: > Is the accrual logic the same here? If so you can call the function from in the task. Or just call it outside the if/else block since it happens in either case.
The payment accrual logic is same, the task only adds the logic to approve a completed work if it has no flags. I will simplify this code to run the update payment accrued function only once and do the completed work approval in the `if` block.

> Relatedly, why does this need to be a separate task? The upload is already happening in a celery task.
I checked the import does not happen in a task only the export for files is done through a celery task.

File: commcare_connect/form_receiver/processor.py
Comment (pxwxnvermx) [reply commcare_connect/form_receiver/processor.py:221]: The bulk processing for the payment approval makes sense. I will change the code to run this task periodically every day using celery schedules.

File: commcare_connect/opportunity/migrations/0041_bulk_approve_completed_work.py
Comment (calellowitz) [inline commcare_connect/opportunity/migrations/0041_bulk_approve_completed_work.py:10]: This will be right in the middle of the workday (12 UTC), It would be safer to do this overnight, maybe midnight UTC or a little earlier like 11pm)

File: commcare_connect/opportunity/tasks.py
Comment (calellowitz) [inline commcare_connect/opportunity/tasks.py:299]: shouldn't this check if the setting to auto approve is enabled for that opportunity?

**Review by calellowitz — COMMENTED** [review]
couple questions but looking better

File: commcare_connect/opportunity/tasks.py
Comment (pxwxnvermx) [reply commcare_connect/opportunity/tasks.py:300]: Fixed. https://github.com/dimagi/commcare-connect/pull/294/commits/7635284b54b8c16e0bd12415bdc2ce54997fd4d9

File: commcare_connect/opportunity/migrations/0041_bulk_approve_completed_work.py
Comment (pxwxnvermx) [reply commcare_connect/opportunity/migrations/0041_bulk_approve_completed_work.py]: https://github.com/dimagi/commcare-connect/pull/294/commits/3c2e32a1d3ec70733ff492b69c687f5c3daf6fe6

**Review by calellowitz — APPROVED** [review]
looks good once the conflict is fixed

**Review by sravfeyn — APPROVED** [review]

**Pull request merged**
